link-grammar/connectors.h

```diff
@@ -198,4 +309,92 @@ static inline unsigned int pair_hash(unsigned int table_size,
 
 	return i & (table_size-1);
 }
+
+static inline condesc_t **condesc_find(ConTable *ct, const char *constring, int hash)
+{
+	size_t i = hash & (ct->size-1);
+
+	while ((NULL != ct->hdesc[i]) &&
+	       !string_set_cmp(constring, ct->hdesc[i]->string))
+	{
+		i = (i + 1) & (ct->size-1);
+	}
+
+	return &ct->hdesc[i];
+}
+
+static inline void condesc_table_alloc(ConTable *ct, size_t size)
+{
+	ct->hdesc = (condesc_t **)malloc(size * sizeof(condesc_t *));
+	memset(ct->hdesc, 0, size * sizeof(condesc_t *));
+	ct->size = size;
+}
+
+static inline bool condesc_insert(ConTable *ct, condesc_t **h,
+                                  const char *constring, int hash)
+{
+	*h = (condesc_t *)malloc(sizeof(condesc_t));
+	memset(*h, 0, sizeof(condesc_t));
```

Comment:
I'm sort-of surprised by this malloc. Why not just change `condesc_table_alloc` so that it mallocs an array of `condesc_t` instead of an array of pointers to `condesc_t` ?  That way, you can skip this smaller alloc entirely.

Reply:
I don't know in advance how many different connectors are in the dict, so this hash table can grow. I didn't want to pre-allocate a an array that cannot grow.

However, I now see it can be done in another,  more efficient, way:
Make a big malloc for the initial guessed number of connectors, as set in dictionary_six_str().
If exceeded, realloc double space. Use it as a `condesc_t` pool. Optionally, at the end, realloc to the exact used amount, in order to free the extra space. Another possibility is allocate by groups of `condesc_t` elements, say 256 and in that way absolutely most of the malloc overhead is saved.

The above realloc way is simpler.  I can implement it.

Reply:
sorry, I did not realize it was a hash table.  It seemed to be getting used as a sequential list.  I suppose I misread the code

Reply:
Um, hang on. `condesc_t` doesn't have a `next` pointer in it, so it cannot be chained, so it cannot be used in a hash table, unless you can guarantee zero collisions..... A hash table with zero-or-one items in each bucket ... Hmmm.  Are you using some trick to spill to the next empty bucket? There was some old code that did that.

Reply:
OK, so you are saying that condesc_t is large enough that it is better to have an array of pointers to them, rather than an array of them?

Reply:
Drawbacks of not using array of pointers:
1. The size of condesc_t is 32, so a table of say 20K items (i think this what it may end up for "ru") will have a significant memory overhead. Since it is too large for an L2 CPU cache, searching in it may cause several times more CPU cache misses (with 8 pointers per cache line you may normally have no more than 3 misses).
2. The sorting, which is essential here, is more costly (need to swap the elements content during sort).

However, I didn't made a benchmark that compares the different ways of doing it.

There is an additional way of doing it that maybe worth exploring: Use the dict creating code.
The idea is just to put the connector strings into their own dict, and enumerate them by listing this dict words (similarly to what `!!*`command does).

Reply:
OK, that's OK. You can leave this alone; I was just reading through the code, got a little confused, and needed to ask a few questions. but this seems OK as is,  No change needed.

On a different note, the `tests/memleak` unit test crashes with a memory double-free. That's important.